docs/architecture/adr-011-light-client-crates-extraction.md

```diff
@@ -0,0 +1,375 @@
+light client implementation would possibly need into a separate crate, say `ibc-base`, and then have both the `ibc`
+modules crate and light client implementation crates depend on that `ibc-base` crate. This was implemented in PR
+[#2327](https://github.com/informalsystems/ibc-rs/pull/2327), but the solution seems to be fragile and unmaintainable.
+
+This ADR proposes to break the circular dependency problem by removing the `Any*` enums completely and using
+trait-objects in their place.
+
+## Decision
+
+### Object safety
+
+There are a total of 5 light client traits, each associated with an `Any*` enum that implements it ->
+
+* `AnyClient` implements `ClientDef`.
+* `AnyClientState` implements `ClientState`.
+* `AnyConsensusState` implements `ConsensusState`.
+* `AnyHeader` implements `Header`.
+* `AnyMisbehaviour` implements `Misbehaviour`.
+
+In order to replace the `Any*` enums with trait objects, these light client traits must be object safe. This
+essentially means that these traits cannot have a `Self: Sized` requirement and their methods cannot have type
+parameters & cannot use `Self`. These restrictions, their implications, and possible workarounds are documented below.
+
+#### Traits that require `Self: Sized` cannot be used as supertraits
+
+Common traits that the light client traits depend on that have the `Self: Sized` requirement
+are `core::convert::{Into<T>, From<T>, ...}`, `tendermint_proto::Protobuf`, `Clone`, etc.
+
+An easy workaround is to add these methods directly to the trait, for e.g. we add an `encode_vec()` method to
+the `Header` trait directly.
+
+```rust
+pub trait ClientState {
+    /// Encode to canonical binary representation
+    fn encode_vec(&self) -> Vec<u8>;
+
+    /* ... */
+}
+/// instead of ->
+/// pub trait ClientState: Protobuf<Any> { /* ... */}
+```
+
+#### `Clone` cannot be a supertrait and cannot be derived for types containing boxed trait objects
+
+The following can be made to work using the `dyn-clone` crate ->
+
+```rust
+#[derive(Clone)]
+pub struct Result {
+    pub client_id: ClientId,
+    pub client_state: Box<dyn ClientState>,
+    pub consensus_state: Box<dyn ConsensusState>,
+}
+```
+
+```rust
+pub trait ClientState: dyn_clone::DynClone { /* ... */ }
+
+dyn_clone::clone_trait_object!(ClientState);
+```
+
+`dyn-clone` also provides a `clone_box()` function that can be used to get a `Box<T>` from a `&T`.
+
+#### Deriving `serde::{Serialize, Deserialize}` for trait objects
+
+Replacing the `AnyHeader` with a `Box<dyn Header>` wouldn't work in the code below because `Deserialize`
+requires `Self: Sized` and `Serialize` uses generic types.
+
+```rust
+#[derive(Deserialize, Serialize)]
+pub struct UpdateClient {
+    pub common: Attributes,
+    pub header: Option<AnyHeader>,
+}
+```
+
+This can be solved using the `ibc_proto::google::protobuf::Any` type instead and having the light client traits provide
+an encoding to `Any` ->
```

Comment:
Instead of using `Option<Any>` for `header`, why not implement `Deserialize` ourselves, and apply the transformation `encode_any()` in `deserialize()`? 

`serde` might also even provide an attribute that we could tag `header` with that specifies to apply this transformation before deserializing. Not sure though.

Reply:
I think the problem with that approach is that the deserialize impl would depend on `Header::decode()` which must have a `Self: Sized` requirement. This is because we must be aware of the concrete type of the header at the time of deserialization but that's not possible.

Reply:
Seems like there are some ready-made solutions like [erased_serde](https://github.com/dtolnay/erased-serde) but I am not sure if they can work for our specific use case.

Reply:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=de9aa123fcec49b410fec8bb23d5f53e

Reply:
> This is because we must be aware of the concrete type of the header at the time of deserialization but that's not possible.

This is a good point. To address this issue, we could modify the `IntoRef` trait [suggested here](https://github.com/informalsystems/ibc-rs/pull/2356/files/619e34127c43c7372af0756a846bea2fef230dc7#r916169288) to inject some context into the conversion:
```rust
trait IntoRef<T: ?Sized> { 
    // Same applies to `try_into_ref()`
    fn into_ref(&self, ctx: &ConversionContext) -> Box<T>;
}
```
`ConversionContext` (or any context more appropriate) would be implemented by the chain (along with all other `Reader`s and `Keeper`s). It would look something like:
```rust
trait ConversionContext {
    fn decode_header(&self, raw_header: &Any) -> Result<Box<dyn Header>, Error>;
    fn decode_client_state(&self, raw_cs: &Any) -> Result<Box<dyn ClientState>, Error>;
    fn decode_consensus_state(&self, raw_cs: &Any) -> Result<Box<dyn ConsensusState>, Error>;
    ...
}
```
Note that we could facilitate the implementation of this trait with the light client registry crate proposed in this ADR (for all known light clients known to the registry).

However, the downside of this approach then is that the `ibc` library doesn't know how to convert `Raw` -> `Domain` for types who use, for example, `dyn Header`. This is because by design, we want the chains to tell us how to decode the `Header` (and `ClientState`, etc) for the light clients that it knows about. 

 We then need to introduce an additional trait `TryFromContext`:
```rust
trait TryFromContext<T> {
    type Error;
    
    fn try_from_ctx(value: T, ctx: ConversionContext) -> Result<Self, Self::Error>;
}
```


which would be used with, say, `MsgUpdateClient`:

```rust
struct MsgUpdateClient {
    pub client_id: ClientId,
    pub header: Box<dyn Header>,
    pub signer: Signer,
}

impl TryFromContext<RawMsgUpdateClient> for MsgUpdateClient {
    type Error = Error;

    // We need to use `TryFromContext` instead of `TryFrom` so that 
    // the `ConversionContext` can be used to decode the raw `Header`
    fn try_from_ctx(raw: RawMsgUpdateClient, ctx: &ConversionContext) -> Result<Self, Self::Error> {
        Ok(MsgUpdateClient {
            client_id: ...,
            header: ctx.decode_header(&raw)?,
            signer: ...,
        })
    }
```

Ultimately, I still prefer this solution to using the raw `Any` for `Header` et al, as it allows us to have proper domain types (where all fields are also domain types), at the cost of more complexity in the conversion code.

WDYT?

Reply:
`erased_serde` uses a cool trick explained here -> https://github.com/dtolnay/erased-serde/blob/master/explanation/main.rs. It allows one to have a separate object safe version of a trait that can operate with the original (non object safe version) seamlessly. 
In our case, I don't think we need to support interoperability with the `tendermint_proto::Protobuf` trait so we can define our own trait that is object safe and just use that instead. We can still use `erased_serde`'s blanket impl trick to keep the original `tendermint_proto::Protobuf` API on the implementation side. 
See experimental PR where `Protobuf` is object safe and requires no other changes on the implementation side -> https://github.com/informalsystems/ibc-rs/pull/2412/files#diff-2b95cbfaf034ca5aa2b9a3f02436e8bd7a7a937df7c5ffb6dba81a0237141941R70 =>
```rust
mod erased {
  use core::convert::{Into as CoreInto, TryFrom as CoreTryFrom};
  
  mod sealed {
      use super::*;
  
      pub trait SealedInto<T: ?Sized> {}
      impl<T, U: Clone + CoreInto<T>> SealedInto<T> for U {}
      pub trait SealedTryFrom<T> {}
      impl<T, U: CoreTryFrom<T>> SealedTryFrom<T> for U {}
  }
  
  pub trait Into<T: ?Sized>: sealed::SealedInto<T> {
      fn into(&self) -> Box<T>;
  }
  
  impl<T, U: Clone + CoreInto<T>> Into<T> for U {
      fn into(&self) -> Box<T> {
          Box::new(self.clone().into())
      }
  }
  
  pub trait TryFrom<T>: sealed::SealedTryFrom<T> {
      type Error;
  
      fn try_from(t: T) -> Result<Self, Self::Error>
      where
          Self: Sized;
  }
  
  impl<T, U: CoreTryFrom<T>> TryFrom<T> for U {
      type Error = <Self as CoreTryFrom<T>>::Error;
  
      fn try_from(t: T) -> Result<Self, Self::Error>
      where
          Self: Sized,
      {
          <Self as CoreTryFrom<T>>::try_from(t)
      }
  }
}

pub trait Protobuf<T: Message + Default>
where
    Self: erased::TryFrom<T> + erased::Into<T>,
    <Self as erased::TryFrom<T>>::Error: Display,
{ /* ... */ }
```

Reply:
`TryFromContext` isn't object safe because the trait function is static (and doesn't have a `self` parameter) and it returns `Self`. This is the same problem we faced with the `From` trait which we solved using the `IntoRef` solution, but that doesn't work here with `TryFromContext`, that's because an equivalent `TryIntoContext` trait bound would look like `T: TryIntoContext<Self>` and that isn't allowed due object safety restrictions. 

Furthermore, to be able to decode using `TryFromContext` we would need to specify the additional context at the point of decoding which may or may not be accessible there, for e.g. Rust's serde deserialization is stateless and so it is not straightforward to access host context at the time of deserialization. 

For these reasons, I believe it is better to not use the `TryFromContext` trait and instead stick to the `TryFrom` trait with a `where Self: Sized` bound.